### ISSUE TITLE
Function expressions

### ISSUE THREAD

**Issue by kengorab** [kengorab/abra-lang#142]
There should be some way to create function expressions, like  lambdas to pass to future `.map`/`.filter`/`.reduce` methods.

This should probably refactor some of the changes made in #141, specifically the `is_anon` field on `TypedFunctionDeclNode`.

**Comment by kengorab** [comment]
This also includes function type identifiers like in `val fn: (String) => String = a => a` for example. This will also require the addition of the `Unit` type, most likely, since there's no way to accurately represent `(a: String) => println(a)`.

**Comment by kengorab** [comment]
In doing this, I should also probably implement optional-safe invocations (and indexing, why not?). That way, the following code will work correctly:
```
val fns: ((String) => String)[] = [
  a => a,
  a => a + "!"
]
println(fns[1]?.("hello")) // Prints "hello!"
println(fns[2]?.("hello")) // Prints None
```

**Comment by kengorab** [comment]
Though this issue is not _fully_ resolved by #158, I'm going to mark this issue as closed. There are a _lot_ more things that came up as a result of doing this work (see #165 and #166) which are kind of blockers for this being done anyway. And rather than keeping this long-lived branch around even longer than it already is, I'm just going to merge it. In, most cases, lambda expressions are working as intended, with the exception of a few edge/corner cases.
